sys/drv/bcm2835_rootdev.c

```diff
@@ -213,52 +110,56 @@ DEVCLASS_DECLARE(emmc);
 static int rootdev_attach(device_t *bus) {
   rootdev_t *rd = bus->state;
 
+  bus->node = FDT_finddevice("/soc/local_intc");
+  if (bus->node == FDT_NODEV)
+    return ENXIO;
+
   rman_init(&rd->rm, "ARM and BCM2835 space");
   rman_manage_region(&rd->rm, BCM2835_PERIPHERALS_BASE,
                      BCM2835_PERIPHERALS_SIZE);
   rman_manage_region(&rd->rm, BCM2836_ARM_LOCAL_BASE, BCM2836_ARM_LOCAL_SIZE);
 
-  rman_init(&rd->irq_rm, "BCM2835 interrupts");
-  rman_manage_region(&rd->irq_rm, 0, NIRQ);
+  rman_init(&rd->irq_rm, "BCM2835 local interrupts");
+  rman_manage_region(&rd->irq_rm, 0, BCM2836_NIRQ);
 
-  /* Map BCM2836 shared processor only once. */
-  rootdev_local_handle = kmem_map_contig(BCM2836_ARM_LOCAL_BASE,
-                                         BCM2836_ARM_LOCAL_SIZE, PMAP_NOCACHE);
-
-  rootdev_arm_base =
-    kmem_map_contig(BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_ARM_BASE),
-                    BCM2835_ARM_SIZE, PMAP_NOCACHE);
+  rd->mem_local = (resource_t){
+    .r_type = RT_MEMORY,
+    .r_bus_tag = generic_bus_space,
+  };
+  int err = bus_space_map(generic_bus_space, BCM2836_ARM_LOCAL_BASE,
+                          BCM2836_ARM_LOCAL_SIZE, &rd->mem_local.r_bus_handle);
+  if (err)
+    return ENXIO;
 
   intr_root_claim(rootdev_intr_handler, bus);
 
-  device_t *dev;
+  /*
+   * Device enumeration.
+   * TODO: this should be performed by a simplebus enumeration.
+   */
+
+  int unit = 0;
+  device_t *bcm2835_pic, *sdhci;
+
+  if ((err = simplebus_add_child(bus, "/soc/intc", unit++, bus, &bcm2835_pic)))
+    return err;
 
-  /* Create ARM timer device and assign resources to it. */
-  dev = device_add_child(bus, 0);
-  dev->pic = bus;
-  device_add_irq(dev, 0, BCM2836_INT_CNTPNSIRQ_CPUN(0));
+  if ((err = simplebus_add_child(bus, "/timer", unit++, bus, NULL)))
+    return err;
 
-  /* Create PL011 UART device and assign resources to it. */
-  dev = device_add_child(bus, 1);
-  dev->pic = bus;
-  device_add_memory(dev, 0, BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_UART0_BASE),
-                    BCM2835_UART0_SIZE);
-  device_add_irq(dev, 0, BCM2835_INT_UART0);
+  if ((err =
+         simplebus_add_child(bus, "/soc/serial", unit++, bcm2835_pic, NULL)))
+    return err;
 
-  /* EMMC */
-  dev = device_add_child(bus, 3);
-  dev->pic = bus;
-  dev->devclass = &DEVCLASS(emmc);
+  if ((err =
+         simplebus_add_child(bus, "/soc/sdhci", unit++, bcm2835_pic, &sdhci)))
+    return err;
```

Comment:
Ok but be consistent. Previously in this function you have pattern:
```
error = foo(...);
if (error)
  return error_code;
```
here you have
```
if ((error = foo(...)))
  return error_code;
```
both of them are readable, but please be consistent.
@cahirwpz we need clean rule for what is allowed.

Reply:
The first one (in line 132) is different because `err` is undefined at that point. But I agree. It's inconsistent and we need a clear convention for such constructs.